sa/metrics.go

```diff
@@ -0,0 +1,125 @@
+package sa
+
+import (
+	"database/sql"
+	"time"
+
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/go-gorp/gorp.v2"
+)
+
+// dbMetrics is a struct holding prometheus stats related to the dbMap. Each of
+// the prometheus stats corresponds to a field of sql.DBStats.
+type dbMetrics struct {
+	dbMap              *gorp.DbMap
+	maxOpenConnections prometheus.Gauge
+	openConnections    prometheus.Gauge
+	inUse              prometheus.Gauge
+	idle               prometheus.Gauge
+	waitCount          prometheus.Counter
+	waitDuration       prometheus.Histogram
+	maxIdleClosed      prometheus.Counter
+	maxLifetimeClosed  prometheus.Counter
+}
+
+// InitDBMetrics will register prometheus stats for the provided dbMap under the
+// given metrics.Scope. Every 1 second in a separate go routine the prometheus
+// stats will be updated based on the gorp dbMap's inner sql.DBMap's DBStats
+// structure values.
+func InitDBMetrics(dbMap *gorp.DbMap, scope metrics.Scope) {
+	// Create a dbMetrics instance and register prometheus metrics
+	dbm := newDbMetrics(dbMap, scope)
+
+	// Start the metric reporting goroutine to update the metrics periodically.
+	go dbm.reportDBMetrics()
+}
+
+// newDbMetrics constructs a dbMetrics instance by registering prometheus stats.
+func newDbMetrics(dbMap *gorp.DbMap, scope metrics.Scope) *dbMetrics {
+	maxOpenConns := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "max_open_connections",
+		Help: "Maximum number of DB connections allowed.",
+	})
+	scope.MustRegister(maxOpenConns)
+
+	openConns := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "open_connections",
+		Help: "Number of established DB connections (in-use and idle).",
+	})
+	scope.MustRegister(openConns)
+
+	inUse := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "inuse",
```

Comment:
Let's give these all a common prefix, like `db_`.

Reply:
Sounds good. Fixed in 9d7d62f